### PR TITLE
Lazily start IncrementingUuidGenerator sessions

### LINKED ISSUE #2930: 7.20.0 regression: TestEngine with ID 'cucumber' failed to discover tests
### 👓 What did you see?

Build of PR opened by renovate with new Cucumber version is failing: https://github.com/jhipster/jhipster-lite/pull/11057:
>  Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.0:test (default-test) on project jhlite: 
Error:  
Error:  Please refer to /home/runner/work/jhipster-lite/jhipster-lite/target/surefire-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
Error: [ERROR] There was an error in the forked process
Error:  TestEngine with ID 'cucumber' failed to discover tests
Error:  org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
Error:  TestEngine with ID 'cucumber' failed to discover tests
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:628)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:285)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:250)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1337)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1137)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:971)
Error:  	at com.gradle.maven.cache.extension.d.c.a(SourceFile:27)

### ✅ What did you expect to see?

Build running fine, as it was with 7.19.0

### 📦 Which tool/library version are you using?

cucumber-junit-platform-engine
cucumber-spring
maven-failsafe-plugin
gradle-develocity

### 🔬 How could we reproduce it?

_No response_

### 📚 Any additional context?

Our cucumbers tests are executed with maven-failsafe-plugin, but here the error happens with the execution of maven-surefire-plugin

### PR DESCRIPTION
**Pull request by mpkorstanje**
### ⚡️ What's your motivation? 

Even when not used, the `IncrementingUuidGenerator` is instantiated with each Cucumber execution. This somewhat fundamental to the way the `ServiceLoader` mechanism works.

Each time an incrementing generator is created, a new session is started for that generator. This means that after 255 executions all sessions are exhausted.

Unfortunately, when using the JUnit Platform, Maven issues a discovery request for each individual class. And Cucumber typically participates in discovery along with JUnit Jupiter. So after requests for 255 classes, Cucumber fails discovery as seen in #2930.

Fixes #2930, #2925

### 🤔 What's changed?

Lazily start IncrementingUuidGenerator sessions

### 🏷️ What kind of change is this?
- :bug: Bug fix (non-breaking change which fixes a defect)

### ♻️ Anything particular you want feedback on?

@jkronegg best I could think of. Got any other ideas?

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://github.com/cucumber/.github/tree/main?tab=coc-ov-file)
- [x] I've changed the behaviour of the code
  - [X] I have added/updated tests to cover my changes.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] Users should know about my change
  - [x] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request.

----

### CODE REVIEW

**Comment by jkronegg** [comment]
**TLDR**: I suggest adding a default `UuidGenerator.initialize()` method which does nothing, but is implemented by `IncrementingUuidGenerator`, and add some test-cases. I can update the PR if you want.

---
The issue #2925 can reproduced easily in `UuidGeneratorServiceLoaderTest` with two different patterns:
- many calls to the `loadUuidGenerator()` method on one `UuidGeneratorServiceLoader` instance
- calling once `loadUuidGenerator()` on many `UuidGeneratorServiceLoader` instance

This corresponds to the two following unit tests:

    @Test
    void test_case_2_no_exception_when_calling_loadUuidGenerator_many_times() {
        Options options = () -> null;
        UuidGeneratorServiceLoader loader = new UuidGeneratorServiceLoader(
                UuidGeneratorServiceLoaderTest.class::getClassLoader,
                options);
        UuidGenerator actual = null;
        for (int i=0; i<1000; i++) {
            actual = loader.loadUuidGenerator();
        }
        assertThat(actual, instanceOf(RandomUuidGenerator.class));
    }

    @Test
    void test_case_2_no_exception_when_instantiating_UuidGeneratorServiceLoader_many_times() {
        Options options = () -> null;
        UuidGenerator actual = null;
        for (int i=0; i<1000; i++) {
            UuidGeneratorServiceLoader loader = new UuidGeneratorServiceLoader(
                    UuidGeneratorServiceLoaderTest.class::getClassLoader,
                    options);
            actual = loader.loadUuidGenerator();
        }
        assertThat(actual, instanceOf(RandomUuidGenerator.class));
    }

>  This somewhat fundamental to the way the `ServiceLoader` mechanism works.

I think it's more related to the logic used to select the UUID generator. When calling the `UuidGeneratorServiceLoader.loadUuidGenerator` the generator will be selected [based on the logic summarized in the test file](https://github.com/cucumber/cucumber-jvm/blob/main/cucumber-core/src/test/java/io/cucumber/core/runtime/UuidGeneratorServiceLoaderTest.java#L21-L36):

> | #   | uuid-generator property   | Available services                                                                  | Result                                                                           |
> |-----|---------------------------|-------------------------------------------------------------------------------------|----------------------------------------------------------------------------------|
> | 1   | undefined                 | none                                                                                | exception, no generators available                                               |
> | 2   | undefined                 | RandomUuidGenerator, IncrementingUuidGenerator                                      | RandomUuidGenerator used                                                         |
> | 3   | RandomUuidGenerator       | RandomUuidGenerator, IncrementingUuidGenerator                                      | RandomUuidGenerator used                                                         |
> | 4   | undefined                 | RandomUuidGenerator, IncrementingUuidGenerator, OtherGenerator                      | OtherGenerator used                                                              |
> | 5   | RandomUuidGenerator       | RandomUuidGenerator, IncrementingUuidGenerator, OtherGenerator                      | RandomUuidGenerator used                                                         |
> | 6   | undefined                 | RandomUuidGenerator, IncrementingUuidGenerator, OtherGenerator, YetAnotherGenerator | exception, cucumber couldn't  decide multiple (non default) generators available |
> | 7   | OtherGenerator            | RandomUuidGenerator, IncrementingUuidGenerator, OtherGenerator, YetAnotherGenerator | OtherGenerator used                                                              |
> | 8   | IncrementingUuidGenerator | RandomUuidGenerator, IncrementingUuidGenerator, OtherGenerator, YetAnotherGenerator | IncrementingUuidGenerator used                                                   |
> | 9   | IncrementingUuidGenerator | RandomUuidGenerator, IncrementingUuidGenerator                                      | IncrementingUuidGenerator used                                                   |
> | 10  | OtherGenerator            | none                                                                                | exception, generator OtherGenerator not available                                |
> | 11  | undefined                 | OtherGenerator                                                                      | OtherGenerator used                                                              |
> | 12  | undefined                 | IncrementingUuidGenerator, OtherGenerator                                           | OtherGenerator used                                                              |
> | 13  | undefined                 | IncrementingUuidGenerator                                                           | IncrementingUuidGenerator used                                                   |

In other words, a custom generator written outside Cucumber has the priority over the `RandomUuidGenerator` and the `IncrementingUuidGenerator`, even if the `cucumber.uuid-generator` property is not specified. To implement this behavior, it is mandatory to iterate over the `ServiceLoader<UuidGenerator>.iterator()`, thus to instantiate the each generator. I don't think we should change this logic (even if the UUID generator is not very popular within the Cucumber community, changing behavior is risky).

Iterating the `ServiceLoader` more than once could be avoided by using caching, but we would need to take a lot of parameters into account (discard the cache when the `cucumber.uuid-generator` property changes or when the classloader changes or loads new classes). But this would be very complex and error-prone due to the many different existing usages of Cucumber.

Lazy loading is the only simple solution that I can see (see the Hyrum's law ;-) ), but I think it should be done only once and not on every call to `generateId()` (because it would decrease the performance). Instead, I suggest introducing a default `initialize()` method which is called only for the selected generator. Something like:

    public interface UuidGenerator extends Supplier<UUID> {
        UUID generateId();

        default UUID get() {
            return generateId();
        }
    
       default void initialize() {
       }
    }

In the `UuidGeneratorServiceLoader`: 

    ....
    UuidGenerator selectedUuidGenerator = ...// decide which generator to use
    selectedUuidGenerator.initialize();
    ...

In the `IncrementingUuidGenerator`:

    public class IncrementingUuidGenerator implements UuidGenerator {
        ...
        public void initialize() {
            initializeMsb();
        }
    }

**Comment by mpkorstanje** [comment]
I think this will do for now.

The ideal solution would be to have the `UuidGenerator` return a supplier instance that keeps the state internally. 

```
public interface UuidGenerator {

    Supplier<UUID> supplier();
}
```

And then have the `UuidGeneratorServiceLoader` return that supplier only for the selected generator.  Fortunately the `UuidGenerator` is experimental so I don't feel too bad about breaking the contract. But I'd rather avoid doing that on a patch version.

**Comment by francoissey** [comment]
I have just upgrade to 7.20.1 and I run into this issue, could we please reopen it?

**Comment by jkronegg** [comment]
@francoissey it seems that you are experiencing another bug variant on this mechanism. Could you please open a new issue to describe your setup and behaviour ? (BTW, this thread is a merged Pull Request: it's not supposed to be reopened)

**Comment by Theo-Hafsaoui** [comment]
Hello, I might be experiencing an issue caused by this regression. Reproducing it is however cumbersome due to the heavy deployment process, and I was wondering if you could estimate which version this bug was introduced in?

**Comment by mpkorstanje** [comment]
Please start a new issue. Otherwise we got nothing to work off.

And be sure to use the `cucumber-bom` to enforce dependency alignment as well as `mvn dependency:tree -Dverbose` to verify alignment.
